text_sampling/text_sampling.py

```diff
@@ -0,0 +1,271 @@
+
+    def __call__(self):
+        sample = self.sampler(self.data)
+        if isinstance(sample, list):  # unwrap list
+            assert len(sample) == 1
+            return sample[0]
+        else:
+            return sample
+
+
+def load_yaml(path):
+    with open(path, "r") as stream:
+        try:
+            data = yaml.safe_load(stream)
+        except yaml.YAMLError as exc:
+            print(exc)
+    return data
+
+
+def get_input_variables_from_template(template: str):
+    return [x for x in re.findall(r"\{([^}]+)\}", template)]
+
+
+class PromptTemplate:
+    def __init__(self, template, input_variables=None):
+        self.template = template
+        if input_variables is None:
+            self.input_variables = self.get_input_variables()
+
+    def get_input_variables(self):
+        return get_input_variables_from_template(self.template)
+
+    def __repr__(self):
+        return f"PromptTemplate: {self.template}"
+
+    def insert(self, data):
+        # check that we got the data for the input_variables, more doesn't matter
+        assert all(x in data.keys() for x in self.input_variables)
+        template = self.template
+        for k in data:
+            template = template.replace("{" + k + "}", data[k])
+        return template
+
+
+def get_random_text_identifiers_and_targets(meta):
+    rnd_texts = {}
+    for e in meta["identifiers"] + meta["targets"]:
+        rnd_texts[e["id"]] = {}
+        if "names" in e:
+            rnd_texts[e["id"]]["names"] = {}
+            name_types = set([list(x.keys())[0] for x in e["names"]])
+            for name in name_types:
+                rnd_text = RandomVariable(
+                    f"{e['id']}__names__{name}",
+                    [x[name] for x in e["names"] if name in x],
+                )
+                rnd_texts[e["id"]]["names"][name] = rnd_text
+        else:
+            rnd_texts[e["id"]]["description"] = partial(
+                lambda x: x, e["description"]
+            )  # to wrap value in function = deterministic, no sampling
+
+    return rnd_texts
+
+
+def get_target_from_string(meta, string):
+    keys = string.split("__")
+
+    def get_with_nested_keys(d, keys):
+        t = d.copy()
+        for k in keys:
+            t = t[k]
+        return t
+
+    if len(keys) == 1 and keys in meta:
+        return meta[
+            keys
+        ]  # assumes single element in meta and doesn't not support nested dicts in that version
+    elif keys[0] in [x["id"] for x in meta["identifiers"] + meta["targets"]]:
+        rnd_texts = get_random_text_identifiers_and_targets(meta)
+        return get_with_nested_keys(rnd_texts, keys)
+    else:
+        raise NotImplementedError()
+
+
+class TemplateSampler:
+    def __init__(
+        self, path_data_dir, template_sampler=None, benchmarking_templates=False
+    ):
+        # TODOS:
+        # lookup table handling, e.g., 3d files for molecules, must come from yaml?!
```

Comment:
yes, then the identifiers (i.e. structures are read from the filepath in the csv)

Reply:
Thank you, I added this info to the to do section in the top comment.